File: react-ios/src/services/jsonImportService.ts
import { getDatabase } from '@/db/index';
import { generateId } from '@/utils/id';

export class JsonImportError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'JsonImportError';
  }
}

export interface ImportResult {
  plansImported: number;
  plansSkipped: number;
  sessionsImported: number;
  sessionsSkipped: number;
  gymsImported: number;
  gymsSkipped: number;
}

export function importResultSummary(result: ImportResult): string {
  const parts: string[] = [];
  if (result.plansImported > 0) parts.push(`${result.plansImported} plans imported`);
  if (result.plansSkipped > 0) parts.push(`${result.plansSkipped} plans skipped (duplicates)`);
  if (result.sessionsImported > 0) parts.push(`${result.sessionsImported} sessions imported`);
  if (result.sessionsSkipped > 0) parts.push(`${result.sessionsSkipped} sessions skipped (duplicates)`);
  if (result.gymsImported > 0) parts.push(`${result.gymsImported} gyms imported`);
  if (result.gymsSkipped > 0) parts.push(`${result.gymsSkipped} gyms skipped (duplicates)`);
  return parts.length === 0 ? 'No data to import.' : parts.join('\n');
}

/**
 * Preview what a unified JSON import contains without importing.
 */
export function previewUnifiedJson(jsonString: string): {
  planCount: number;
  sessionCount: number;
  gymCount: number;
  hasSettings: boolean;
} {
  const json = parseJsonString(jsonString);
  const plans = Array.isArray(json.plans) ? json.plans : [];
  const sessions = Array.isArray(json.sessions) ? json.sessions : [];
  const singleSession = json.session && typeof json.session === 'object' ? 1 : 0;
  const gyms = Array.isArray(json.gyms) ? json.gyms : [];
  const hasSettings = json.settings != null && typeof json.settings === 'object'
    && Object.keys(json.settings as Record<string, unknown>).length > 0;

  return {
    planCount: plans.length,
    sessionCount: sessions.length + singleSession,
    gymCount: gyms.length,
    hasSettings,
  };
}

/**
 * Validate and import a unified JSON string.
 * Uses merge semantics: skips duplicates by name+date for sessions, name for plans, name for gyms.
 */
export async function importUnifiedJson(jsonString: string): Promise<ImportResult> {
  const json = parseJsonString(jsonString);

  // Check format version if present
  if (json.formatVersion && json.formatVersion !== '1.0') {
    throw new JsonImportError(`Unsupported format version: ${json.formatVersion}`);
  }

  const result: ImportResult = {
    plansImported: 0,
    plansSkipped: 0,
    sessionsImported: 0,
    sessionsSkipped: 0,
    gymsImported: 0,
    gymsSkipped: 0,
  };

  const db = await getDatabase();

  await db.execAsync('BEGIN TRANSACTION');

  try {
    // Import plans
    if (Array.isArray(json.plans)) {
      for (const planData of json.plans) {
        await importPlan(planData as Record<string, unknown>, result);
      }
    }

    // Import sessions (array format)
    if (Array.isArray(json.sessions)) {
      for (const sessionData of json.sessions) {
        await importSession(sessionData as Record<string, unknown>, result);
      }
    }

    // Also handle single session format (from single-session exports)
    if (json.session && typeof json.session === 'object' && !Array.isArray(json.session)) {
      await importSession(json.session as Record<string, unknown>, result);
    }

    // Import gyms
    if (Array.isArray(json.gyms)) {
      for (const gymData of json.gyms) {
        await importGym(gymData as Record<string, unknown>, result);
      }
    }

    await db.execAsync('COMMIT');
  } catch (error) {
    await db.execAsync('ROLLBACK');
    throw error;
  }

  return result;
}

// ============================================================================
// Private Helpers
// ============================================================================

function parseJsonString(jsonString: string): Record<string, unknown> {
  let json: unknown;
  try {
    json = JSON.parse(jsonString);
  } catch {
    throw new JsonImportError('File is not valid JSON.');
  }

  if (!json || typeof json !== 'object' || Array.isArray(json)) {
    throw new JsonImportError('File is not a valid JSON object.');
  }

  const obj = json as Record<string, unknown>;

  // Must have at least one data section
  if (!obj.plans && !obj.sessions && !obj.session) {
    throw new JsonImportError('File does not contain any importable data (no plans, sessions, or session found).');
  }

  return obj;
}

async function importPlan(
  data: Record<string, unknown>,
  result: ImportResult
): Promise<void> {
  const name = data.name as string | undefined;
  if (!name) return;

  const db = await getDatabase();

  // Check for duplicate by name
  const existing = await db.getFirstAsync<{ count: number }>(
    'SELECT COUNT(*) as count FROM workout_templates WHERE name = ?',
    [name]
  );
  if (existing && existing.count > 0) {
    result.plansSkipped += 1;
    return;
  }

  const planId = generateId();
  const now = new Date().toISOString();
  const tags = Array.isArray(data.tags) ? data.tags : [];
  const tagsJson = JSON.stringify(tags);

  await db.runAsync(
    `INSERT INTO workout_templates (id, name, description, tags, default_weight_unit, source_markdown, created_at, updated_at, is_favorite)
     VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)`,
    [
      planId,
      name,
      (data.description as string) || null,
      tagsJson,
      (data.defaultWeightUnit as string) || null,
      (data.sourceMarkdown as string) || null,
      now,
      now,
      data.isFavorite === true ? 1 : 0,
    ]
  );

  // Import exercises
  const exercises = Array.isArray(data.exercises) ? data.exercises : [];
  for (const exerciseData of exercises) {
    const ex = exerciseData as Record<string, unknown>;
    const exerciseId = generateId();

    await db.runAsync(
      `INSERT INTO template_exercises (id, workout_template_id, exercise_name, order_index, notes, equipment_type, group_type, group_name, parent_exercise_id)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)`,
      [
        exerciseId,
        planId,
        (ex.exerciseName as string) || 'Unknown',
        (ex.orderIndex as number) ?? 0,
        (ex.notes as string) || null,
        (ex.equipmentType as string) || null,
        (ex.groupType as string) || null,
        (ex.groupName as string) || null,
        null, // parent_exercise_id
      ]
    );

    // Import sets — note: template_sets lacks is_amrap and notes columns
    const sets = Array.isArray(ex.sets) ? ex.sets : [];
    for (const setData of sets) {
      const s = setData as Record<string, unknown>;
      await db.runAsync(
        `INSERT INTO template_sets (id, template_exercise_id, order_index, target_weight, target_weight_unit, target_reps, target_time, target_rpe, rest_seconds, tempo, is_dropset, is_per_side)
         VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
        [
          generateId(),
          exerciseId,
          (s.orderIndex as number) ?? 0,
          (s.targetWeight as number) ?? null,
          (s.targetWeightUnit as string) || null,
          (s.targetReps as number) ?? null,
          (s.targetTime as number) ?? null,
          (s.targetRpe as number) ?? null,
          (s.restSeconds as number) ?? null,
          (s.tempo as string) || null,
          s.isDropset === true ? 1 : 0,
          s.isPerSide === true ? 1 : 0,
        ]
      );
    }
  }

  result.plansImported += 1;
}

async function importSession(
  data: Record<string, unknown>,
  result: ImportResult
): Promise<void> {
  const name = data.name as string | undefined;
  const date = data.date as string | undefined;
  if (!name || !date) return;

  const db = await getDatabase();

  // Check for duplicate by name + date
  const existing = await db.getFirstAsync<{ count: number }>(
    'SELECT COUNT(*) as count FROM workout_sessions WHERE name = ? AND date = ?',
    [name, date]
  );
  if (existing && existing.count > 0) {
    result.sessionsSkipped += 1;
    return;
  }

  const sessionId = generateId();

  await db.runAsync(
    `INSERT INTO workout_sessions (id, workout_template_id, name, date, start_time, end_time, duration, notes, status)
     VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)`,
    [
      sessionId,
      null, // no template association
      name,
      date,
      (data.startTime as string) || null,
      (data.endTime as string) || null,
      (data.duration as number) ?? null,
      (data.notes as string) || null,
      (data.status as string) || 'completed',
    ]
  );

  // Import exercises
  const exercises = Array.isArray(data.exercises) ? data.exercises : [];
  for (const exerciseData of exercises) {
    const ex = exerciseData as Record<string, unknown>;
    const exerciseId = generateId();

    await db.runAsync(
      `INSERT INTO session_exercises (id, workout_session_id, exercise_name, order_index, notes, equipment_type, group_type, group_name, parent_exercise_id, status)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
      [
        exerciseId,
        sessionId,
        (ex.exerciseName as string) || 'Unknown',
        (ex.orderIndex as number) ?? 0,
        (ex.notes as string) || null,
        (ex.equipmentType as string) || null,
        (ex.groupType as string) || null,
        (ex.groupName as string) || null,
        null, // parent_exercise_id
        (ex.status as string) || 'completed',
      ]
    );

    // Import sets
    const sets = Array.isArray(ex.sets) ? ex.sets : [];
    for (const setData of sets) {
      const s = setData as Record<string, unknown>;
      await db.runAsync(
        `INSERT INTO session_sets (id, session_exercise_id, order_index, parent_set_id, drop_sequence,
         target_weight, target_weight_unit, target_reps, target_time, target_rpe, rest_seconds,
         actual_weight, actual_weight_unit, actual_reps, actual_time, actual_rpe,
         completed_at, status, notes, tempo, is_dropset, is_per_side)
         VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
        [
          generateId(),
          exerciseId,
          (s.orderIndex as number) ?? 0,
          null, // parent_set_id
          null, // drop_sequence
          (s.targetWeight as number) ?? null,
          (s.targetWeightUnit as string) || null,
          (s.targetReps as number) ?? null,
          (s.targetTime as number) ?? null,
          (s.targetRpe as number) ?? null,
          (s.restSeconds as number) ?? null,
          (s.actualWeight as number) ?? null,
          (s.actualWeightUnit as string) || null,
          (s.actualReps as number) ?? null,
          (s.actualTime as number) ?? null,
          (s.actualRpe as number) ?? null,
          (s.completedAt as string) || null,
          (s.status as string) || 'completed',
          (s.notes as string) || null,
          (s.tempo as string) || null,
          s.isDropset === true ? 1 : 0,
          s.isPerSide === true ? 1 : 0,
        ]
      );
    }
  }

  result.sessionsImported += 1;
}

async function importGym(
  data: Record<string, unknown>,
  result: ImportResult
): Promise<void> {
  const name = data.name as string | undefined;
  if (!name) return;

  const db = await getDatabase();

  // Check for duplicate by name
  const existing = await db.getFirstAsync<{ count: number }>(
    'SELECT COUNT(*) as count FROM gyms WHERE name = ?',
    [name]
  );
  if (existing && existing.count > 0) {
    result.gymsSkipped += 1;
    return;
  }

  const now = new Date().toISOString();
  await db.runAsync(
    `INSERT INTO gyms (id, name, is_default, created_at, updated_at)
     VALUES (?, ?, ?, ?, ?)`,
    [
      generateId(),
      name,
      0, // Not default
      now,
      now,
    ]
  );

  result.gymsImported += 1;
}

File: mobile-apps/ios/LiftMark/Services/JsonImportService.swift
import Foundation
import GRDB

/// Errors thrown during JSON import operations.
enum JsonImportError: LocalizedError {
    case invalidFormat(String)
    case unsupportedVersion(String)
    case importFailed(String)

    var errorDescription: String? {
        switch self {
        case .invalidFormat(let reason):
            return "Invalid import file: \(reason)"
        case .unsupportedVersion(let version):
            return "Unsupported format version: \(version)"
        case .importFailed(let reason):
            return "Import failed: \(reason)"
        }
    }
}

/// Service for importing unified JSON exports into the database.
/// Uses merge semantics: skips duplicates by name+date for sessions, name for plans.
struct JsonImportService {

    struct ImportResult {
        var plansImported: Int = 0
        var plansSkipped: Int = 0
        var sessionsImported: Int = 0
        var sessionsSkipped: Int = 0
        var gymsImported: Int = 0
        var gymsSkipped: Int = 0

        var summary: String {
            var parts: [String] = []
            if plansImported > 0 { parts.append("\(plansImported) plans imported") }
            if plansSkipped > 0 { parts.append("\(plansSkipped) plans skipped (duplicates)") }
            if sessionsImported > 0 { parts.append("\(sessionsImported) sessions imported") }
            if sessionsSkipped > 0 { parts.append("\(sessionsSkipped) sessions skipped (duplicates)") }
            if gymsImported > 0 { parts.append("\(gymsImported) gyms imported") }
            if gymsSkipped > 0 { parts.append("\(gymsSkipped) gyms skipped (duplicates)") }
            return parts.isEmpty ? "No data to import." : parts.joined(separator: "\n")
        }
    }

    /// Validate and import a unified JSON file.
    func importUnifiedJson(from url: URL) throws -> ImportResult {
        let data = try Data(contentsOf: url)
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw JsonImportError.invalidFormat("File is not valid JSON.")
        }

        // Check format version if present
        if let version = json["formatVersion"] as? String, version != "1.0" {
            throw JsonImportError.unsupportedVersion(version)
        }

        var result = ImportResult()

        try DatabaseManager.shared.database().write { db in
            // Import plans
            if let plans = json["plans"] as? [[String: Any]] {
                for planData in plans {
                    try importPlan(planData, into: db, result: &result)
                }
            }

            // Import sessions
            if let sessions = json["sessions"] as? [[String: Any]] {
                for sessionData in sessions {
                    try importSession(sessionData, into: db, result: &result)
                }
            }

            // Also handle single session format (from single-session exports)
            if let sessionData = json["session"] as? [String: Any] {
                try importSession(sessionData, into: db, result: &result)
            }

            // Import gyms
            if let gyms = json["gyms"] as? [[String: Any]] {
                for gymData in gyms {
                    try importGym(gymData, into: db, result: &result)
                }
            }
        }

        return result
    }

    /// Check if a file at the given URL is a valid LiftMark JSON export.
    func validateJsonFile(at url: URL) -> Bool {
        guard let data = try? Data(contentsOf: url),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return false
        }
        // Must have at least one of: plans, sessions, session
        return json["plans"] != nil || json["sessions"] != nil || json["session"] != nil
    }

    // MARK: - Private Import Helpers

    private func importPlan(_ data: [String: Any], into db: Database, result: inout ImportResult) throws {
        guard let name = data["name"] as? String else { return }

        // Check for duplicate by name
        let existingCount = try Int.fetchOne(db, sql: "SELECT COUNT(*) FROM workout_templates WHERE name = ?", arguments: [name]) ?? 0
        if existingCount > 0 {
            result.plansSkipped += 1
            return
        }

        let planId = UUID().uuidString
        let now = ISO8601DateFormatter().string(from: Date())
        let tags = (data["tags"] as? [String]) ?? []
        let tagsJson = (try? JSONSerialization.data(withJSONObject: tags))
            .flatMap { String(data: $0, encoding: .utf8) } ?? "[]"

        try db.execute(sql: """
            INSERT INTO workout_templates (id, name, description, tags, default_weight_unit, source_markdown, created_at, updated_at, is_favorite)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
            """, arguments: [
                planId,
                name,
                data["description"] as? String,
                tagsJson,
                data["defaultWeightUnit"] as? String,
                data["sourceMarkdown"] as? String,
                now,
                now,
                (data["isFavorite"] as? Bool) == true ? 1 : 0
            ])

        // Import exercises
        if let exercises = data["exercises"] as? [[String: Any]] {
            // First pass: create exercises and track ID mapping for parent references
            var exerciseIdMap: [Int: String] = [:] // orderIndex -> new ID
            for exerciseData in exercises {
                let exerciseId = UUID().uuidString
                let orderIndex = exerciseData["orderIndex"] as? Int ?? 0
                exerciseIdMap[orderIndex] = exerciseId

                try db.execute(sql: """
                    INSERT INTO template_exercises (id, workout_template_id, exercise_name, order_index, notes, equipment_type, group_type, group_name, parent_exercise_id)
                    VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
                    """, arguments: [
                        exerciseId,
                        planId,
                        exerciseData["exerciseName"] as? String ?? "Unknown",
                        orderIndex,
                        exerciseData["notes"] as? String,
                        exerciseData["equipmentType"] as? String,
                        exerciseData["groupType"] as? String,
                        exerciseData["groupName"] as? String,
                        nil as String? // parent resolved in second pass
                    ])

                // Import sets
                if let sets = exerciseData["sets"] as? [[String: Any]] {
                    for setData in sets {
                        try db.execute(sql: """
                            INSERT INTO template_sets (id, template_exercise_id, order_index,
                                target_weight, target_weight_unit, target_reps, target_time,
                                target_rpe, rest_seconds, tempo, is_dropset, is_per_side,
                                is_amrap, notes)
                            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                            """, arguments: [
                                UUID().uuidString,
                                exerciseId,
                                setData["orderIndex"] as? Int ?? 0,
                                setData["targetWeight"] as? Double,
                                setData["targetWeightUnit"] as? String,
                                setData["targetReps"] as? Int,
                                setData["targetTime"] as? Int,
                                setData["targetRpe"] as? Int,
                                setData["restSeconds"] as? Int,
                                setData["tempo"] as? String,
                                (setData["isDropset"] as? Bool) == true ? 1 : 0,
                                (setData["isPerSide"] as? Bool) == true ? 1 : 0,
                                (setData["isAmrap"] as? Bool) == true ? 1 : 0,
                                setData["notes"] as? String
                            ])
                    }
                }
            }
        }

        result.plansImported += 1
    }

    private func importSession(_ data: [String: Any], into db: Database, result: inout ImportResult) throws {
        guard let name = data["name"] as? String,
              let date = data["date"] as? String else { return }

        // Check for duplicate by name + date
        let existingCount = try Int.fetchOne(db, sql: "SELECT COUNT(*) FROM workout_sessions WHERE name = ? AND date = ?", arguments: [name, date]) ?? 0
        if existingCount > 0 {
            result.sessionsSkipped += 1
            return
        }

        let sessionId = UUID().uuidString

        try db.execute(sql: """
            INSERT INTO workout_sessions (id, workout_template_id, name, date, start_time, end_time, duration, notes, status)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
            """, arguments: [
                sessionId,
                nil as String?,
                name,
                date,
                data["startTime"] as? String,
                data["endTime"] as? String,
                data["duration"] as? Int,
                data["notes"] as? String,
                data["status"] as? String ?? "completed"
            ])

        // Import exercises
        if let exercises = data["exercises"] as? [[String: Any]] {
            for exerciseData in exercises {
                let exerciseId = UUID().uuidString

                try db.execute(sql: """
                    INSERT INTO session_exercises (id, workout_session_id, exercise_name, order_index, notes, equipment_type, group_type, group_name, parent_exercise_id)
                    VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
                    """, arguments: [
                        exerciseId,
                        sessionId,
                        exerciseData["exerciseName"] as? String ?? "Unknown",
                        exerciseData["orderIndex"] as? Int ?? 0,
                        exerciseData["notes"] as? String,
                        exerciseData["equipmentType"] as? String,
                        exerciseData["groupType"] as? String,
                        exerciseData["groupName"] as? String,
                        nil as String?
                    ])

                // Import sets
                if let sets = exerciseData["sets"] as? [[String: Any]] {
                    for setData in sets {
                        try db.execute(sql: """
                            INSERT INTO session_sets (id, session_exercise_id, order_index,
                                parent_set_id, drop_sequence, target_weight,
                                target_weight_unit, target_reps, target_time, target_rpe,
                                rest_seconds, actual_weight, actual_weight_unit, actual_reps,
                                actual_time, actual_rpe, completed_at, status, notes, tempo,
                                is_dropset, is_per_side)
                            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                            """, arguments: [
                                UUID().uuidString,
                                exerciseId,
                                setData["orderIndex"] as? Int ?? 0,
                                nil as String?, // parent_set_id
                                nil as Int?,    // drop_sequence
                                setData["targetWeight"] as? Double,
                                setData["targetWeightUnit"] as? String,
                                setData["targetReps"] as? Int,
                                setData["targetTime"] as? Int,
                                setData["targetRpe"] as? Int,
                                setData["restSeconds"] as? Int,
                                setData["actualWeight"] as? Double,
                                setData["actualWeightUnit"] as? String,
                                setData["actualReps"] as? Int,
                                setData["actualTime"] as? Int,
                                setData["actualRpe"] as? Int,
                                setData["completedAt"] as? String,
                                setData["status"] as? String ?? "completed",
                                setData["notes"] as? String,
                                setData["tempo"] as? String,
                                (setData["isDropset"] as? Bool) == true ? 1 : 0,
                                (setData["isPerSide"] as? Bool) == true ? 1 : 0
                            ])
                    }
                }
            }
        }

        result.sessionsImported += 1
    }

    private func importGym(_ data: [String: Any], into db: Database, result: inout ImportResult) throws {
        guard let name = data["name"] as? String else { return }

        // Check for duplicate by name
        let existingCount = try Int.fetchOne(db, sql: "SELECT COUNT(*) FROM gyms WHERE name = ?", arguments: [name]) ?? 0
        if existingCount > 0 {
            result.gymsSkipped += 1
            return
        }

        let now = ISO8601DateFormatter().string(from: Date())
        try db.execute(sql: """
            INSERT INTO gyms (id, name, is_default, created_at, updated_at)
            VALUES (?, ?, ?, ?, ?)
            """, arguments: [
                UUID().uuidString,
                name,
                0, // Not default
                now,
                now
            ])

        result.gymsImported += 1
    }
}
